Optimize random tick logic in SpreadingSnowyDirtBlock

WhiteProject1 opened the pull request:
Refactored random tick code to use primitive coordinates for performance, added early skip for zero offsets, and delayed BlockPos creation. Improved chunk access and block state checks to reduce unnecessary object creation and method calls, enhancing efficiency during grass and dirt spreading.

WhiteProject1 committed:
Optimize random tick logic in SpreadingSnowyDirtBlock

Refactored random tick code to use primitive coordinates for performance, added early skip for zero offsets, and delayed BlockPos creation. Improved chunk access and block state checks to reduce unnecessary object creation and method calls, enhancing efficiency during grass and dirt spreading.

R00tB33rMan reviewed (approved):
Unless I'm blind, this LGTM. I can't really find anything that'd incur any discrepancies of concern

coredex-source reviewed (approved):
Looks alright to me.

Dueris commented on canvas-server/minecraft-patches/sources/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java.patch:
```diff
@@ -9,37 +9,56 @@
++                    final int offsetX = random.nextInt(3) - 1;
++                    final int offsetY = random.nextInt(5) - 3;
++                    final int offsetZ = random.nextInt(3) - 1;
++                    if (offsetX == 0 && offsetY == 0 && offsetZ == 0) { // Canvas - optimize random tick (early skip)
```
Just a minor nitpick, since we are already in a `Canvas start`/`Canvas end` block, don't really need the `Canvas - optimize random tick` part here. Can just be `// early skip`

Dueris commented on canvas-server/minecraft-patches/sources/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java.patch:
```diff
@@ -9,37 +9,56 @@
-+                        // Canvas end - optimize random tick
+-                    if (access.getBlockStateFinal(blockPos.getX(), blockPos.getY(), blockPos.getZ()).is(Blocks.DIRT) && SpreadingSnowyDirtBlock.canPropagate(access, blockState, level, blockPos)) {
+-                        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, blockPos, blockState.setValue(SNOWY, isSnowySetting(access.getBlockStateFinal(blockPos.getX(), blockPos.getY() + 1, blockPos.getZ()))), 3); // CraftBukkit
++                    final BlockState blockStateAtPos = access.getBlockStateFinal(blockX, blockY, blockZ); // Canvas - optimize random tick
```
Same thing here in these few bits, we are already in a `Canvas start`/`Canvas end` block. This would only really apply if the change intent was different, but since all are `optimize random tick`, I don't see a point in having those commented

Dueris reviewed (changes requested):
Beyond the minor nitpicks, this looks good to me

Dueris commented:
This patch also doesn't apply correctly, the hunk fails. Please fix this.

WhiteProject1 committed:
Refactor snow/dirt spreading for performance

Optimized random tick logic in SpreadingSnowyDirtBlock by reducing unnecessary BlockPos creation and chunk/block state lookups. Early checks and chunk access improvements help reduce overhead during block spread events.

Dueris commented:
Still cannot apply patches.

WhiteProject1 committed:
hunk

hunk

Dueris commented:
This still does not apply.

Dueris commented:
I am closing this. Please learn the patch system first and make a proper patch that *does* apply, compile, and run correctly before PRing.
